mlflow/R/mlflow/R/databricks-utils.R

```diff
@@ -98,27 +98,62 @@ get_databricks_config <- function(profile) {
   config
 }
 
+#' @return A list of tags to be set by the run context when creating MLflow runs in the
+#' current Databricks Notebook environment
+build_context_tags_from_databricks_notebook_info <- function(notebook_info) {
+  tags <- list()
+  tags[[MLFLOW_DATABRICKS_TAGS$MLFLOW_DATABRICKS_NOTEBOOK_ID]] <- notebook_info$id
+  tags[[MLFLOW_DATABRICKS_TAGS$MLFLOW_DATABRICKS_NOTEBOOK_PATH]] <- notebook_info$path
+  tags[[MLFLOW_DATABRICKS_TAGS$MLFLOW_DATABRICKS_WEBAPP_URL]] <- notebook_info$webapp_url
+  tags[[MLFLOW_TAGS$MLFLOW_SOURCE_NAME]] <- notebook_info$path
+  tags[[MLFLOW_TAGS$MLFLOW_SOURCE_VERSION]] <- get_source_version()
+  tags[[MLFLOW_TAGS$MLFLOW_SOURCE_TYPE]] <- MLFLOW_SOURCE_TYPE$NOTEBOOK
+  tags
+}
+
+#' @return A list of tags to be set by the run context when creating MLflow runs in the
+#' current Databricks Job environment
+build_context_tags_from_databricks_job_info <- function(job_info) {
+  tags <- list()
+  tags[[MLFLOW_DATABRICKS_TAGS$MLFLOW_DATABRICKS_JOB_ID]] <- job_info$job_id
+  tags[[MLFLOW_DATABRICKS_TAGS$MLFLOW_DATABRICKS_JOB_RUN_ID]] <- job_info$run_id
+  tags[[MLFLOW_DATABRICKS_TAGS$MLFLOW_DATABRICKS_JOB_TYPE]] <- job_info$job_type
+  tags[[MLFLOW_DATABRICKS_TAGS$MLFLOW_DATABRICKS_WEBAPP_URL]] <- job_info$webapp_url
+  tags[[MLFLOW_TAGS$MLFLOW_SOURCE_NAME]] <- paste(
+    "jobs", job_info$job_id, "run", job_info$run_id, sep = "/"
+  )
+  tags[[MLFLOW_TAGS$MLFLOW_SOURCE_VERSION]] <- get_source_version()
+  tags[[MLFLOW_TAGS$MLFLOW_SOURCE_TYPE]] <- MLFLOW_SOURCE_TYPE$JOB
+  tags
+}
+
 mlflow_get_run_context.mlflow_databricks_client <- function(client, experiment_id, ...) {
   if (exists(".databricks_internals")) {
-    notebook_info <- do.call(".get_notebook_info", list(), envir = get(".databricks_internals",
-                                                                       envir = .GlobalEnv))
+    databricks_internal_env <- get(".databricks_internals", envir = .GlobalEnv)
+    notebook_info <- do.call(".get_notebook_info", list(), envir = databricks_internal_env)
     if (!is.na(notebook_info$id) && !is.na(notebook_info$path)) {
-      tags <- list()
-      tags[[MLFLOW_DATABRICKS_TAGS$MLFLOW_DATABRICKS_NOTEBOOK_ID]] <- notebook_info$id
-      tags[[MLFLOW_DATABRICKS_TAGS$MLFLOW_DATABRICKS_NOTEBOOK_PATH]] <- notebook_info$path
-      tags[[MLFLOW_DATABRICKS_TAGS$MLFLOW_DATABRICKS_WEBAPP_URL]] <- notebook_info$webapp_url
-      tags[[MLFLOW_TAGS$MLFLOW_SOURCE_NAME]] <- notebook_info$path
-      tags[[MLFLOW_TAGS$MLFLOW_SOURCE_VERSION]] <- get_source_version()
-      tags[[MLFLOW_TAGS$MLFLOW_SOURCE_TYPE]] <- MLFLOW_SOURCE_TYPE$NOTEBOOK
-      list(
+      return(list(
         client = client,
-        tags = tags,
+        tags = build_context_tags_from_databricks_notebook_info(notebook_info),
         experiment_id = experiment_id %||% notebook_info$id,
         ...
-      )
+      ))
```

Comment:
Would it be cleaner to do this as if(notebook) {...} elseif (job) {...} else {nextMethod()} ?

Reply:
This would look cleaner, but I think we'd have to fetch the job info prior to the if/else-if/else block to clean this up, resulting in extra function call within Databricks notebooks. Let me know if you see a way around this / think that's a better idea than the current solution.

Reply:
Oh I see. I think that's ok then.